### ISSUE TITLE
Switch from `Proposer` to `Slots`

### ISSUE THREAD

**Issue by bkchr** [paritytech/cumulus#36]
Instead of using `Proposer` as interface for building the blocks in the collator, we probably want to switch over to `Slots` as the interface. At the point when we want to support custom authoring, using `Slots` is easier as Babe and Aura already implemented using this trait.

We would need to change the trait in Substrate to return the `Proposal`, because we need this in the collator.

**Comment by JoshOrndorff** [comment]
You mean switching to [`SlotWorker`](https://crates.parity.io/sc_consensus_slots/trait.SlotWorker.html) right? Or maybe [`SimpleSlotWorker`](https://crates.parity.io/sc_consensus_slots/trait.SimpleSlotWorker.html)?

Would this change eliminate the possibility of using PoW for collating? Is that a compromise we really want?

**Comment by bkchr** [comment]
> You mean switching to [`SlotWorker`](https://crates.parity.io/sc_consensus_slots/trait.SlotWorker.html) right? Or maybe [`SimpleSlotWorker`](https://crates.parity.io/sc_consensus_slots/trait.SimpleSlotWorker.html)?

`SlotWorker`.

> 
> Would this change eliminate the possibility of using PoW for collating? Is that a compromise we really want?

We don't really support PoW at the moment at all. So, yes.

**Comment by JoshOrndorff** [comment]
I started looking at making the change to the `SlotWorker` trait, and I have a few questions about it.

1. [`Proposal`](https://crates.parity.io/sp_consensus/struct.Proposal.html) is generic over the `Transaction` type used in the backend. Should I make the `SlotWorker` trait also generic over this type?

2. Currently the only implementation of the `SlotWorker` trait come from the [blanket implementation](https://github.com/paritytech/substrate/blob/26f3d6826a181674257b8e9fc708aa04b49dc73c/client/consensus/slots/src/lib.rs#L385-L391) over the `SimpleSlotWorker` trait.
That means that I'll need to update the provided `on_slot` method for the `SimpleSlotWorker` trait. I'm stuck because I can't figure out how to return the proposal [here](https://github.com/paritytech/substrate/blob/26f3d6826a181674257b8e9fc708aa04b49dc73c/client/consensus/slots/src/lib.rs#L378). After it's data has already been moved [eg here](https://github.com/paritytech/substrate/blob/26f3d6826a181674257b8e9fc708aa04b49dc73c/client/consensus/slots/src/lib.rs#L345). It seems Proposal doesn't implement `Clone`.

3. I see you mentioned on the Babe issue that it wouldn't happen soon. I'm curious if that is true for slots in general, or just for babe. (My main goal right now is Aura)
